File: init_db.py
"""
Run once to create all tables and load seed data.
Usage: python init_db.py
"""
import os
import pymysql
from dotenv import load_dotenv
from app import create_app
from app.extensions import db

load_dotenv()


def run_sql_file(cursor, path):
    with open(path) as f:
        sql = f.read()
    for statement in sql.split(";"):
        stmt = statement.strip()
        if stmt:
            cursor.execute(stmt)


def main():
    app = create_app()
    with app.app_context():
        print("Creating tables...")
        db.create_all()
        print("Tables created.")

        conn = pymysql.connect(
            host=os.environ["DB_HOST"],
            port=int(os.environ.get("DB_PORT", 3306)),
            user=os.environ["DB_USER"],
            password=os.environ["DB_PASSWORD"],
            database=os.environ["DB_NAME"],
            charset="utf8mb4",
        )
        cursor = conn.cursor()

        seeds_dir = os.path.join(os.path.dirname(__file__), "database", "seeds")
        seed_files = [
            "naics_codes.sql",
            "keywords.sql",
            "user_settings.sql",
            "funders.sql",
            "data_sources.sql",
        ]

        for fname in seed_files:
            path = os.path.join(seeds_dir, fname)
            if os.path.exists(path):
                print(f"  Loading {fname}...")
                try:
                    run_sql_file(cursor, path)
                    conn.commit()
                except Exception as e:
                    print(f"  Warning: {e} (may already exist)")
                    conn.rollback()

        cursor.close()
        conn.close()
        print("Done. Database is ready.")
        print("\nLog in with:")
        print(f"  Username: {os.environ.get('ADMIN_USERNAME', 'aisha')}")
        print(f"  Password: [from your .env ADMIN_PASSWORD]")


if __name__ == "__main__":
    main()

File: init_db.js
require('dotenv').config();
const mysql = require('mysql2/promise');
const fs = require('fs');
const path = require('path');
const { dedupeFunders } = require('./database/migrations/dedupe_funders');
const { upgradeBidWritingGuides } = require('./database/migrations/upgrade_bid_writing_guides');
const { seedBidWritingGuides } = require('./database/seeds/bid_writing_guides');
const { updateNaicsCodes } = require('./database/migrations/update_naics_codes');
const { dedupeDataSources } = require('./database/migrations/dedupe_data_sources');
const { scoreAllUnscored } = require('./services/alignmentScorer');

async function init() {
  const conn = await mysql.createConnection({
    host: process.env.DB_HOST || 'localhost',
    port: process.env.DB_PORT || 3306,
    database: process.env.DB_NAME,
    user: process.env.DB_USER,
    password: process.env.DB_PASSWORD,
    multipleStatements: true
  });

  console.log('Connected to database.');

  const schema = fs.readFileSync(path.join(__dirname, 'database/schema.sql'), 'utf8');
  await conn.query(schema);
  console.log('Schema applied.');

  const seeds = [
    'database/seeds/naics_codes.sql',
    'database/seeds/keywords.sql',
    'database/seeds/funders.sql',
    'database/seeds/data_sources.sql',
    'database/seeds/user_settings.sql'
  ];

  for (const seedFile of seeds) {
    const sql = fs.readFileSync(path.join(__dirname, seedFile), 'utf8');
    try {
      await conn.query(sql);
      console.log(`Seeded: ${seedFile}`);
    } catch (err) {
      if (err.code === 'ER_DUP_ENTRY') {
        console.log(`Skipped (already seeded): ${seedFile}`);
      } else {
        throw err;
      }
    }
  }

  const removed = await dedupeFunders(conn);
  if (removed > 0) console.log(`Removed ${removed} duplicate funder row(s).`);

  const dataSourcesRemoved = await dedupeDataSources(conn);
  if (dataSourcesRemoved > 0) console.log(`Removed ${dataSourcesRemoved} duplicate data source row(s).`);

  await upgradeBidWritingGuides(conn);
  const guidesAdded = await seedBidWritingGuides(conn);
  console.log(`Bid Writing Guides: ${guidesAdded} added.`);

  const naicsRemoved = await updateNaicsCodes(conn);
  if (naicsRemoved > 0) console.log(`Removed ${naicsRemoved} outdated NAICS code(s).`);

  const scored = await scoreAllUnscored();
  if (scored > 0) console.log(`Alignment score computed for ${scored} previously-unscored opportunit${scored === 1 ? 'y' : 'ies'}.`);

  await conn.end();
  console.log('Database initialized successfully.');
}

init().catch(err => {
  console.error('Init failed:', err.message);
  process.exit(1);
});
